### ISSUE TITLE
cameraSource.release() throws null pointer exception if detector is null

### ISSUE THREAD

**Issue by sudipbhandari126** [googlesamples/android-vision#184]
`java.lang.NullPointerException: Attempt to invoke virtual method 'void com.google.android.gms.vision.Detector.release()' on a null object reference`

According to [https://developers.google.com/android/reference/com/google/android/gms/vision/CameraSource.html#release()](Android Documentation) release method completely shuts down this camera source and releases the resources of the underlying detector. But sometimes detector is null and this method throws exception.

Would it not be better to handle camerasource and detector resource separately in that case?

**Comment by pm0733464** [comment]
This issue only happens if you call CameraSource.release() more than once.  CameraSource assumes that it won't be used after release() is called.

**Comment by vpdsouza** [comment]
Nope, I debugged the code to make sure the release is only called once and it is.  But the moment I restart it I get this: 

`Exception thrown from receiver.
                                                                    java.lang.NullPointerException: Attempt to invoke virtual method 'void com.google.android.gms.vision.Detector.receiveFrame(com.google.android.gms.vision.Frame)' on a null object reference
                                                                        at com.imbills.mobile.barcode.ui.camera.CameraSource$FrameProcessingRunnable.run(CameraSource.java:1205)`

**Comment by LucianIacob** [comment]
Did someone found any way to avoid this crash?

**Comment by Davehp44** [comment]
Did someone found any way to avoid this crash? @LucianIacob

**Comment by LucianIacob** [comment]
@Davehp44 we solved this in a way like:
```
if (mCameraSource != null) {
     try {
         mCameraSource.release();
     } catch (NullPointerException ignored) {  }
     mCameraSource = null;
}
```
And then for starting camera:
```
if (cameraSource == null) {
     releaseCamera();
     return;
}

mCameraSource = cameraSource;
mCameraSource.start(mSurfaceView.getHolder());
```

**Comment by roopeshn28** [comment]
can you tell where exactly u are changing the starting camera?

**Comment by mattdonders** [comment]
I don't mind using the workaround provided by @LucianIacob but is there an actual fix for this code or should we always be checking the state of the camerasource in code?

**Comment by datnt1987** [comment]
So @LucianIacob 's solution is the best choice at this time ???
